File: Trees/000 Binary Tree/Leaf-Similar Trees/RecursiveSolution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    ArrayList<TreeNode> r1Leafs = new ArrayList();
    int pos=0;
    boolean similar = true;
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        getPreOrder(root1);
        checkLeaf(root2);
        if(pos!=r1Leafs.size())
            similar = false;
        return similar;
    }

    public void getPreOrder(TreeNode root) {
        if(root==null)
            return;
        getPreOrder(root.left);
        getPreOrder(root.right);

        if(root.left==null && root.right==null)
            r1Leafs.add(root);
    }

    public void checkLeaf(TreeNode root) {
        if(root==null)
            return;
        checkLeaf(root.left);
        checkLeaf(root.right);
        if(root.left==null && root.right==null) {
            if(pos>=r1Leafs.size()||root.val!=r1Leafs.get(pos++).val)
                similar=false;
        }
    }
}

File: Trees/000 Binary Tree/Leaf-Similar Trees/RecursiveSolution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun leafSimilar(root1: TreeNode?, root2: TreeNode?): Boolean {
        val r1Leaf:ArrayList<Int> = getPostOrder(root1,ArrayList<Int>())
        val r2Leaf:ArrayList<Int> = getPostOrder(root2,ArrayList<Int>())

        return r1Leaf.equals(r2Leaf)
    }

    fun getPostOrder(root: TreeNode?, leaf:ArrayList<Int>):ArrayList<Int> {
        root ?: return leaf
        getPostOrder(root.left,leaf)
        getPostOrder(root.right,leaf)

        if(root.left==null && root.right==null)
            leaf.add(root.`val`)
        
        return leaf
    }
}
